// redux-form/lib/__tests__/helpers/reducer.arrayUnshift.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _actions = _interopRequireDefault(require("../../actions"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var arrayUnshift = _actions.default.arrayUnshift;

var describeArrayUnshift = function describeArrayUnshift(reducer, expect, _ref) {
  var fromJS = _ref.fromJS;
  return function () {
    it('should work with empty state', function () {
      var state = reducer(undefined, arrayUnshift('foo', 'myField', 'myValue'));
      expect(state).toEqualMap({
        foo: {
          values: {
            myField: ['myValue']
          }
        }
      });
    });
    it('should insert at beginning', function () {
      var state = reducer(fromJS({
        foo: {
          values: {
            myField: {
              subField: ['a', 'b', 'c']
            }
          },
          fields: {
            myField: {
              subField: [{
                touched: true
              }, {
                touched: true,
                visited: true
              }, {
                touched: true
              }]
            }
          }
        }
      }), arrayUnshift('foo', 'myField.subField', 'newValue'));
      expect(state).toEqualMap({
        foo: {
          values: {
            myField: {
              subField: ['newValue', 'a', 'b', 'c']
            }
          },
          fields: {
            myField: {
              subField: [{}, {
                touched: true
              }, {
                touched: true,
                visited: true
              }, {
                touched: true
              }]
            }
          }
        }
      });
    });
  };
};

var _default = describeArrayUnshift;
exports.default = _default;
