// redux-form/lib/__tests__/helpers/reducer.arraySplice.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _actions = _interopRequireDefault(require("../../actions"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var arraySplice = _actions.default.arraySplice;

var describeArraySplice = function describeArraySplice(reducer, expect, _ref) {
  var fromJS = _ref.fromJS;
  return function () {
    it('should work with empty state', function () {
      var state = reducer(undefined, arraySplice('foo', 'myField', 0, 0, 'myValue'));
      expect(state).toEqualMap({
        foo: {
          values: {
            myField: ['myValue']
          }
        }
      });
    });
    it('should work with existing form errors', function () {
      var state = reducer(fromJS({
        foo: {
          values: {
            myField: {
              subField: ['a']
            }
          },
          fields: {
            myField: {
              subField: [{
                touched: true
              }]
            }
          },
          submitErrors: {
            myField: {
              subField: ['invalid value']
            }
          },
          asyncErrors: {
            myField: {
              subField: ['invalid format']
            }
          }
        }
      }), arraySplice('foo', 'myField.subField', 0, 0, 'myValue'));
      expect(state).toEqualMap({
        foo: {
          values: {
            myField: {
              subField: ['myValue', 'a']
            }
          },
          fields: {
            myField: {
              subField: [{}, {
                touched: true
              }]
            }
          },
          submitErrors: {
            myField: {
              subField: [undefined, 'invalid value']
            }
          },
          asyncErrors: {
            myField: {
              subField: [undefined, 'invalid format']
            }
          }
        }
      });
    });
    it('should insert at beginning', function () {
      var state = reducer(fromJS({
        foo: {
          values: {
            myField: {
              subField: ['a', 'b', 'c']
            }
          },
          fields: {
            myField: {
              subField: [{
                touched: true
              }, {
                touched: true,
                visited: true
              }, {
                touched: true
              }]
            }
          }
        }
      }), arraySplice('foo', 'myField.subField', 0, 0, 'newValue'));
      expect(state).toEqualMap({
        foo: {
          values: {
            myField: {
              subField: ['newValue', 'a', 'b', 'c']
            }
          },
          fields: {
            myField: {
              subField: [{}, {
                touched: true
              }, {
                touched: true,
                visited: true
              }, {
                touched: true
              }]
            }
          }
        }
      });
    });
    it('should insert at end', function () {
      var state = reducer(fromJS({
        foo: {
          values: {
            myField: {
              subField: ['a', 'b', 'c']
            }
          },
          fields: {
            myField: {
              subField: [{
                touched: true
              }, {
                touched: true,
                visited: true
              }, {
                touched: true
              }]
            }
          }
        }
      }), arraySplice('foo', 'myField.subField', 3, 0, 'newValue'));
      expect(state).toEqualMap({
        foo: {
          values: {
            myField: {
              subField: ['a', 'b', 'c', 'newValue']
            }
          },
          fields: {
            myField: {
              subField: [{
                touched: true
              }, {
                touched: true,
                visited: true
              }, {
                touched: true
              }, {}]
            }
          }
        }
      });
    });
    it('should insert in middle', function () {
      var state = reducer(fromJS({
        foo: {
          values: {
            myField: {
              subField: ['a', 'b', 'c']
            }
          },
          fields: {
            myField: {
              subField: [{
                touched: true
              }, {
                touched: true,
                visited: true
              }, {
                touched: true
              }]
            }
          }
        }
      }), arraySplice('foo', 'myField.subField', 1, 0, 'newValue'));
      expect(state).toEqualMap({
        foo: {
          values: {
            myField: {
              subField: ['a', 'newValue', 'b', 'c']
            }
          },
          fields: {
            myField: {
              subField: [{
                touched: true
              }, {}, {
                touched: true,
                visited: true
              }, {
                touched: true
              }]
            }
          }
        }
      });
    });
    it('should remove from beginning', function () {
      var state = reducer(fromJS({
        foo: {
          values: {
            myField: {
              subField: ['a', 'b', 'c', 'd']
            }
          },
          fields: {
            myField: {
              subField: [{
                touched: true,
                visited: true
              }, {
                touched: true
              }, {
                touched: true,
                visited: true
              }, {
                touched: true
              }]
            }
          }
        }
      }), arraySplice('foo', 'myField.subField', 0, 1));
      expect(state).toEqualMap({
        foo: {
          values: {
            myField: {
              subField: ['b', 'c', 'd']
            }
          },
          fields: {
            myField: {
              subField: [{
                touched: true
              }, {
                touched: true,
                visited: true
              }, {
                touched: true
              }]
            }
          }
        }
      });
    });
    it('should remove from end', function () {
      var state = reducer(fromJS({
        foo: {
          values: {
            myField: {
              subField: ['a', 'b', 'c', 'd']
            }
          },
          fields: {
            myField: {
              subField: [{
                touched: true,
                visited: true
              }, {
                touched: true
              }, {
                touched: true,
                visited: true
              }, {
                touched: true
              }]
            }
          }
        }
      }), arraySplice('foo', 'myField.subField', 3, 1));
      expect(state).toEqualMap({
        foo: {
          values: {
            myField: {
              subField: ['a', 'b', 'c']
            }
          },
          fields: {
            myField: {
              subField: [{
                touched: true,
                visited: true
              }, {
                touched: true
              }, {
                touched: true,
                visited: true
              }]
            }
          }
        }
      });
    });
    it('should remove from middle', function () {
      var state = reducer(fromJS({
        foo: {
          values: {
            myField: {
              subField: ['a', 'b', 'c', 'd']
            }
          },
          fields: {
            myField: {
              subField: [{
                touched: true,
                visited: true
              }, {
                touched: true
              }, {
                touched: true,
                visited: true
              }, {
                touched: true
              }]
            }
          }
        }
      }), arraySplice('foo', 'myField.subField', 1, 1));
      expect(state).toEqualMap({
        foo: {
          values: {
            myField: {
              subField: ['a', 'c', 'd']
            }
          },
          fields: {
            myField: {
              subField: [{
                touched: true,
                visited: true
              }, {
                touched: true,
                visited: true
              }, {
                touched: true
              }]
            }
          }
        }
      });
    });
  };
};

var _default = describeArraySplice;
exports.default = _default;
