// redux-form/lib/__tests__/helpers/reducer.destroy.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _actions = _interopRequireDefault(require("../../actions"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var destroy = _actions.default.destroy;

var describeDestroy = function describeDestroy(reducer, expect, _ref) {
  var fromJS = _ref.fromJS;
  return function () {
    it('should destroy form state', function () {
      var state = reducer(fromJS({
        foo: {
          values: {
            myField: 'initialValue'
          },
          active: 'myField'
        },
        otherThing: {
          touchThis: false
        }
      }), destroy('foo'));
      expect(state).toEqualMap({
        otherThing: {
          touchThis: false
        }
      });
    });
    it('should destroy form state for multiple forms', function () {
      var state = reducer(fromJS({
        foo: {
          values: {
            myField: 'fooInitialValue'
          },
          active: 'myFooField'
        },
        bar: {
          values: {
            myField: 'barInitialValue'
          },
          active: 'myBarField'
        },
        otherThing: {
          touchThis: false
        }
      }), destroy('foo', 'bar'));
      expect(state).toEqualMap({
        otherThing: {
          touchThis: false
        }
      });
    });
  };
};

var _default = describeDestroy;
exports.default = _default;
