// redux-form/es/__tests__/helpers/reducer.clearAsyncError.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _actions = _interopRequireDefault(require("../../actions"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var clearAsyncError = _actions.default.clearAsyncError;

var describeClearAsyncError = function describeClearAsyncError(reducer, expect, _ref) {
  var fromJS = _ref.fromJS;
  return function () {
    it('should do nothing on clear submit with no previous state', function () {
      var state = reducer(undefined, clearAsyncError('foo'));
      expect(state).toEqualMap({
        foo: {}
      });
    });
    it('should clear async errors with previous state', function () {
      var state = reducer(fromJS({
        myForm: {
          asyncErrors: {
            foo: 'some validation message here',
            baar: 'second validation message'
          }
        }
      }), clearAsyncError('myForm', 'foo'));
      expect(state).toEqualMap({
        myForm: {
          asyncErrors: {
            baar: 'second validation message'
          }
        }
      });
    });
  };
};

var _default = describeClearAsyncError;
exports.default = _default;
