// redux-form/es/__tests__/helpers/reducer.arrayPop.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _actions = _interopRequireDefault(require("../../actions"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var arrayPop = _actions.default.arrayPop;

var describeArrayPop = function describeArrayPop(reducer, expect, _ref) {
  var fromJS = _ref.fromJS;
  return function () {
    it('should do nothing with no array', function () {
      var state = reducer(fromJS({
        foo: {}
      }), arrayPop('foo', 'myField.subField'));
      expect(state).toEqualMap({
        foo: {}
      });
    });
    it('should pop from end', function () {
      var state = reducer(fromJS({
        foo: {
          values: {
            myField: {
              subField: ['a', 'b', 'c', 'd']
            }
          },
          fields: {
            myField: {
              subField: [{
                touched: true,
                visited: true
              }, {
                touched: true
              }, {
                touched: true,
                visited: true
              }, {
                touched: true
              }]
            }
          }
        }
      }), arrayPop('foo', 'myField.subField'));
      expect(state).toEqualMap({
        foo: {
          values: {
            myField: {
              subField: ['a', 'b', 'c']
            }
          },
          fields: {
            myField: {
              subField: [{
                touched: true,
                visited: true
              }, {
                touched: true
              }, {
                touched: true,
                visited: true
              }]
            }
          }
        }
      });
    });
  };
};

var _default = describeArrayPop;
exports.default = _default;
