// redux-form/es/__tests__/helpers/reducer.startAsyncValidation.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _actions = _interopRequireDefault(require("../../actions"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var startAsyncValidation = _actions.default.startAsyncValidation;

var describeStartAsyncValidation = function describeStartAsyncValidation(reducer, expect, _ref) {
  var fromJS = _ref.fromJS;
  return function () {
    it('should set asyncValidating on startAsyncValidation', function () {
      var state = reducer(fromJS({
        foo: {
          doesnt: 'matter',
          should: 'notchange'
        }
      }), startAsyncValidation('foo'));
      expect(state).toEqualMap({
        foo: {
          doesnt: 'matter',
          should: 'notchange',
          asyncValidating: true
        }
      });
    });
    it('should set asyncValidating with field name on startAsyncValidation', function () {
      var state = reducer(fromJS({
        foo: {
          values: {
            myField: 'initialValue'
          },
          initial: {
            myField: 'initialValue'
          }
        }
      }), startAsyncValidation('foo', 'myField'));
      expect(state).toEqualMap({
        foo: {
          values: {
            myField: 'initialValue'
          },
          initial: {
            myField: 'initialValue'
          },
          asyncValidating: 'myField'
        }
      });
    });
  };
};

var _default = describeStartAsyncValidation;
exports.default = _default;
