// redux-form/es/__tests__/helpers/reducer.clearSubmitErrors.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _actions = _interopRequireDefault(require("../../actions"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var clearSubmitErrors = _actions.default.clearSubmitErrors;

var describeClearSubmitErrors = function describeClearSubmitErrors(reducer, expect, _ref) {
  var fromJS = _ref.fromJS;
  return function () {
    it('should do nothing on clear submit errors with no previous state', function () {
      var state = reducer(undefined, clearSubmitErrors('foo'));
      expect(state).toEqualMap({
        foo: {}
      });
    });
    it('should clear submit errors with previous state', function () {
      var state = reducer(fromJS({
        foo: {
          values: {
            myField: 'initialValue'
          },
          initial: {
            myField: 'initialValue'
          },
          fields: {
            myField: {},
            otherField: {
              visited: true,
              active: true
            }
          },
          submitErrors: {
            some: 'error'
          },
          error: 'form-wide error',
          active: 'otherField',
          triggerSubmit: true
        }
      }), clearSubmitErrors('foo'));
      expect(state).toEqualMap({
        foo: {
          values: {
            myField: 'initialValue'
          },
          initial: {
            myField: 'initialValue'
          },
          fields: {
            myField: {},
            otherField: {
              visited: true,
              active: true
            }
          },
          active: 'otherField',
          triggerSubmit: true
        }
      });
    });
  };
};

var _default = describeClearSubmitErrors;
exports.default = _default;
